src/SimpleSMT/Solver.hs

```diff
@@ -0,0 +1,383 @@
+
+data Backend = Backend {
+  send :: Builder -> IO LBS.ByteString
+  -- ^ Send a command to the backend.
+  }
+
+type Queue = IORef Builder
+
+-- | Push a command on the solver's queue of commands to evaluate.
+-- The command must not produce any output when evaluated, unless it is the last
+-- command added before the queue is flushed.
+putQueue :: Queue -> SExpr -> IO ()
+putQueue q expr = atomicModifyIORef q $ \cmds ->
+  (cmds <> renderSExpr expr, ())
+
+-- | Empty the queue of commands to evaluate and return its content as a bytestring
+-- builder.
+flushQueue :: Queue -> IO Builder
+flushQueue q = atomicModifyIORef q $ \cmds ->
+    (mempty, cmds)
+
+data Solver = Solver
+  { backend :: Backend
+  -- ^ The backend processing the commands.
+  , queue :: Maybe Queue
+  -- ^ A queue to write commands that are to be sent to the solver lazily.
+  , log :: LBS.ByteString -> IO ()
+  -- ^ The function used for logging the solver's activity.
+  }
+
+sendSolver :: Solver -> Builder -> IO SExpr
+sendSolver solver cmd = do
+  log solver $ "[send] " <> toLazyByteString cmd
+  resp <- send (backend solver) cmd
+  case parseSExpr resp of
+    Nothing -> do
+      log solver $ "[error] failed to parse solver output:\n" <> resp
+      fail $
+        unlines ["Unexpected response from the SMT solver:", "parsing failed"]
+    Just (expr, _) -> do
+      log solver $ "[recv] " <> toLazyByteString (renderSExpr expr)
+      return expr
+
+-- | Create a new solver and initialize it with some options so that it behaves
+-- correctly for our use.
+initSolverWith ::
+     Backend
+  -- | whether to enable lazy mode
+  -> Bool
+  -- | function for logging the solver's activity
+  -> (LBS.ByteString -> IO ())
+  -> IO Solver
+initSolverWith solverBackend lazy logger = do
+  solverQueue <- if lazy then do
+      ref <- newIORef mempty
+      return $ Just ref
+    else return Nothing
+  let solver = Solver solverBackend solverQueue logger
+  if lazy then
+      return ()
+    else
+      -- this should not be enabled when the queue is used, as it messes with parsing
+      -- the outputs of commands that are actually interesting
+      -- TODO checking for correctness and enabling laziness can be made compatible
+      -- but it would require the solver backends to return list of s-expressions
+      -- alternatively, we may consider that the user wanting both features should
+      -- implement their own backend that deals with this
+      setOption solver ":print-success" "true"
+  setOption solver ":produce-models" "true"
+  return solver
+
+
+-- | Have the solver evaluate a command in SExpr format.
+-- This forces the queued commands to be evaluated as well, but their results are
+-- *not* checked for correctness.
+command :: Solver -> SExpr -> IO SExpr
+command solver expr = do
+  let cmd = renderSExpr expr
+  sendSolver solver =<<
+    case queue solver of
+      Nothing -> return $ cmd
+      Just q -> (<> renderSExpr expr) <$> flushQueue q
```

Comment:
How about reusing the logic of queues for commands taking `SExpr` and commands taking `Builder`?

Reply:
I'm not sure I understand what you mean, could you develop ?

Reply:
Sure. As a user of this library, which function should I call to send commands to the SMT solver if I want to have queuing of commands but I want to provide a `Builder` and not a `SExpr`?

Does that clarify it?

Reply:
Yes I get your point. Currently there's no way to do that, as I was assuming the user would only be writing `SExpr`s. But it makes sense to make it possible indeed.